sdks/java/io/kinesis/src/main/java/org/apache/beam/sdk/io/kinesis/KinesisReader.java

```diff
@@ -142,4 +159,28 @@ public Instant getWatermark() {
     return source;
   }
 
+  /**
+   * Returns total size of all records that remain in Kinesis stream after current watermark.
+   * When currently processed record is not further behind than {@link #upToDateThreshold}
+   * then this method returns 0.
+   */
+  @Override
+  public long getTotalBacklogBytes() {
+    Instant watermark = getWatermark();
+    if (watermark.plus(upToDateThreshold).isAfterNow()) {
+      return 0L;
+    }
+    if (backlogBytesLastCheckTime.plus(backlogBytesCheckThreshold).isAfterNow()) {
```

Comment:
is this method called so often that we need to cache the response from CloudWatch?

Reply:
I'm not sure how often this gets called in other runners but it wasn't a problem in case of Dataflow. This was suggested by @rangadi as a precaution.

Reply:
Another important aspect is that comment implied this backlog is updated on Kenesis only at minute granularity. Fetching more often does not have any advantage.

Dataflow calls getWatermark() method after every "bundle" which could take only a few milliseconds or at most on the order of second. Assuming fetching backlog adds latency due to RPC to external system, it could severely affect performance, without providing any benefit.